tests/Integration/Core/Configuration/AbstractMultistoreConfigurationTest.php

```diff
@@ -0,0 +1,238 @@
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Integration\Core\Configuration;
+
+use Configuration as LegacyConfiguration;
+use PrestaShop\PrestaShop\Adapter\Configuration;
+use PrestaShop\PrestaShop\Adapter\Shop\Context;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+use PrestaShop\PrestaShop\Core\Feature\FeatureInterface;
+use PrestaShopBundle\Service\Form\MultistoreCheckboxEnabler;
+use Shop;
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+use Symfony\Component\OptionsResolver\Exception\InvalidOptionsException;
+use Symfony\Component\OptionsResolver\Exception\UndefinedOptionsException;
+use Tests\Resources\DummyMultistoreConfiguration;
+
+class AbstractMultistoreConfigurationTest extends KernelTestCase
+{
+    /**
+     * @var Configuration
+     */
+    protected $legacyConfiguration;
+
+    /**
+     * @var Context
+     */
+    protected $shopContext;
+
+    /**
+     * @var FeatureInterface
+     */
+    protected $multistoreFeature;
+
+    /**
+     * @var Shop
+     */
+    protected $newShop;
+
+    public function setUp(): void
+    {
+        self::bootKernel();
+        $this->entityManager = self::$kernel->getContainer()->get('doctrine.orm.entity_manager');
+        $this->legacyConfiguration = self::$kernel->getContainer()->get('prestashop.adapter.legacy.configuration');
+        $this->initMultistore();
+        $this->multistoreFeature = self::$kernel->getContainer()->get('prestashop.adapter.multistore_feature');
+    }
+
+    /**
+     * @dataProvider provideShopConstraints
+     *
+     * @param ShopConstraint $shopConstraint
+     * @param bool $isAllShopContext
+     */
+    public function testUpdate(ShopConstraint $shopConstraint, bool $isAllShopContext): void
+    {
+        $testedObject = $this->getTestableObject($shopConstraint, $isAllShopContext);
+
+        if ($isAllShopContext) {
+            // in all shop we test without multistore checkboxes (it would throw an exception, this is tested elsewhere)
+            $testedObject->updateConfiguration(['test_conf_1' => true, 'test_conf_2' => 'string_result']);
```

Comment:
`test_conf_1`, `test_conf_2` should come from the dataprovider :thinking: 
You're always testing the same values, whatever which parameters are set, if the updateConfiguration failed in the second dataprovider loop, you'll not see it.

Why not using Behat for this kind of tests? :thinking:

Reply:
The input data should be a combination of values to update for a shop constraint, and a set of values to check by for specific shop constraint

```php
public function dataProvider(): iterable
{
    // First test changes the config for all shops which impacts also shop and shopGroup
    yield [
        // Shop constraint used for update
        ShopConstraint::allShops(),
        // Data for update
        ['test_conf_1' => true, 'test_conf_2' => 'all_shop_conf2'],
        // List of checks to do (for different shop constraints, which implies creating a Configuration object for each one, in a loop)
        [
            [
                ShopConstraint::allShops(),
                ['test_conf_1' => true, 'test_conf_2' => 'all_shop_conf2'],
            ],
            [
                ShopConstraint::shopGroup(1),
                ['test_conf_1' => true, 'test_conf_2' => 'all_shop_conf2'],
            ],
            [
                ShopConstraint::shop(1),
                ['test_conf_1' => true, 'test_conf_2' => 'all_shop_conf2'],
            ],
        ]
    ];

    // Second test changes the config for single shop which does not impact all shops and shopGroup, only one field is checked
    yield [
        ShopConstraint::shop(1),
        ['test_conf_1' => false, 'test_conf_2' => 'single_shop_conf2', MultistoreCheckboxEnabler::MULTISTORE_FIELD_PREFIX . 'test_conf_2' => true],
        [
            [
                ShopConstraint::allShops(),
                ['test_conf_1' => true, 'test_conf_2' => 'all_shop_conf2'],
            ],
            [
                ShopConstraint::shopGroup(1),
                ['test_conf_1' => true, 'test_conf_2' => 'all_shop_conf2'],
            ],
            [
                ShopConstraint::shop(1),
                // Only test_conf_2 is modified since it was the only checkbox enabled
                ['test_conf_1' => true, 'test_conf_2' => 'single_shop_conf2'],
            ],
        ]
    ];
}
```